src/base_object-inl.h

```diff
@@ -95,13 +95,12 @@ void BaseObject::MakeWeak() {
   persistent_handle_.SetWeak(
       this,
       [](const v8::WeakCallbackInfo<BaseObject>& data) {
-        BaseObject* obj = data.GetParameter();
+        std::unique_ptr<BaseObject> obj(data.GetParameter());
         // Clear the persistent handle so that ~BaseObject() doesn't attempt
         // to mess with internal fields, since the JS object may have
         // transitioned into an invalid state.
         // Refs: https://github.com/nodejs/node/issues/18897
         obj->persistent_handle_.Reset();
```

Comment:
So … I’m not exactly sure whether this is something I overlooked when refactoring `BaseObject` or not, but shouldn’t the `if (persistent_handle_.IsEmpty()) return;` in `~BaseObject()` make this part unnecessary?

Reply:
I'll take a closer look at this tomorrow. Thanks

Reply:
My understanding of this is that when the callback/finalizer lambda is called, the underlying `Persistent` object on the heap will have been freed, but the `BaseObject`
instance still has a reference to it via `persistent_handle_`. 
`BaseObject` uses this reference in its destructor to create a new `Local<Persistent>`, and then calls `SetAlignedPointerInInternalField`. This will cause a segment fault when calling `OpenHandle` as the handle has already been freed, unless we call `Reset` that is. Calling `obj->persistent_handle_.Reset()` will set the state of the `persistent_handle_` and set it's value to `0` which we then check in the destructor using `persistent_handle_.IsEmpty()`. 

This is my first time looking closer at how `SetWeak` works so please correct me if I'm wrong and I'd be happy to revisit this next week.

Reply:
No, you’re absolutely right – thanks for verifying this!